content/blog/2021/08/2021-08-20-jenkins-operator-security-work-report.adoc

```diff
@@ -0,0 +1,67 @@
+---
+
+== Background
+
+Jenkins custom resources on a Kubernetes cluster are deployed using declarative YAML configuration files; hence some of the plugins declared in these files may contain security warnings.So there is no way for the user to know other than manually checking for each on the site. 
+This project aims to add an extra step of validation before creating/updating a new Jenkins Custom Resource.
+
+== Deliverables 
+
+This project aims to add a validating admission webhook to the jenkins kubernetes operator to detect potential security vulnerabilities in the plugins before the object is created.
+
+== Dependencies 
+
+Webhooks communicate to the API server over HTTPS and use TLS. Thus, https://cert-manager.io/docs/[Jetstack/cert-manager] is used to provision TLS certificates and establish connection between Kubernetes API and webhook.  
+
+== Implementaion
+
+Operator-SDK takes care of creating a new webhook and appending it to the manager and creating handlers. Tls certificates are managed using cert-manager.
+
+* Validation Logic:
+** Proposed Implementations: Iterate through the list of plugins to be installed and fetch warnings for each plugin from the plugin center API and check if the version of that plugin has any of those warnings.  
+** Caveats: Webhooks add to API request latency,hence they should evaluate as quickly as possible thus having max allowed timeout of 30s. In the earlier approach I was fetching the security warnings from the plugin site API in the validator interface itself, and since network operations slow it was causing a timeout in the case when the number of plugins were more or internet connection was not good.
+** Updated Implementaion: Instead of fetching information for each plugin, the information of all the plugins is downloaded and cached at the start of the operator and updated periodically,thus eliminating network calls and finishing validation in less than a second.
+ 
+=== Evaluation Phase 1: 
+* Scaffoled a new validation webhook
+* Added manifests for ValidatingWebhookConfiguration, certificates and volumes, and updated Makefile
+* Implemented the validator interface
+* Updated helm charts
+
+=== Evaluation Phase 2:
+* Reimplemented the validator interface.
+* Added unit tests for internal functions
+* Added e2e tests along with helm tests 
+* Updated helm charts 
+
+== Resources
+* Pull Requests
+** https://github.com/jenkinsci/kubernetes-operator/pull/585[Added validation webhook,manifests,and updated Makefile]
+** https://github.com/jenkinsci/kubernetes-operator/pull/593[Implemented validation logic,added tests and updated helm charts]
+*  https://www.youtube.com/watch?v=xO2jGmv1fLo[Phase 1 demo]
+
+== User Guide 
+The webhook feature is completely optional for the user. It can be easily deployed it using helm charts by setting webhook.enabled in values.yaml or in the command line flag.
```

Comment:
```suggestion
The webhook feature is completely optional for the user. It can be easily deployed using Helm Chart by setting webhook.enabled in values.yaml and in the Operator command line flag.
```